### PR TITLE
Redis

### PR DESCRIPTION
**Pull request by ajax-marchenko-s**
- add redis repository for artwork
- add tests for redis repository

### CODE REVIEW

File: domainservice/src/main/kotlin/ua/marchenko/artauction/artwork/repository/impl/MongoArtworkRepository.kt [hunk 1]
```diff
@@ -67,12 +67,12 @@ internal class MongoArtworkRepository(
...
 
     override fun updateStatusByIdAndPreviousStatus(
-        artworkId: String,
+        id: String,
         prevStatus: ArtworkStatus,
         newStatus: ArtworkStatus
     ): Mono<MongoArtwork> {
         val query = Query.query(
-            Criteria.where(MongoArtwork::id.name).`is`(artworkId).and(MongoArtwork::status.name).`is`(prevStatus)
+            Criteria.where(MongoArtwork::id.name).`is`(id).and(MongoArtwork::status.name).`is`(prevStatus)
```
Comment (ajax-torbin-v) [inline domainservice/src/main/kotlin/ua/marchenko/artauction/artwork/repository/impl/MongoArtworkRepository.kt:75]: isEqualTo ?

File: domainservice/src/main/kotlin/ua/marchenko/artauction/artwork/repository/impl/RedisArtworkRepository.kt [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+internal class RedisArtworkRepository(
+    private val mongoArtworkRepository: MongoArtworkRepository,
+    private val reactiveRedisTemplate: ReactiveRedisTemplate<String, ByteArray>,
+    private val objectMapper: ObjectMapper,
+) : ArtworkRepository by mongoArtworkRepository {
+
+    override fun save(artwork: MongoArtwork): Mono<MongoArtwork> =
+        mongoArtworkRepository.save(artwork)
+            .flatMap { savedArtwork ->
+                val id = requireNotNull(savedArtwork.id) { "artwork id cannot be null" }.toHexString()
```
Comment (ajax-gnatiuk-s) [inline domainservice/src/main/kotlin/ua/marchenko/artauction/artwork/repository/impl/RedisArtworkRepository.kt:31]: seems, `requireNotNull` is redundant. `id` here cannot be null

File: domainservice/src/main/kotlin/ua/marchenko/artauction/artwork/repository/impl/RedisArtworkRepository.kt [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+                reactiveRedisTemplate.delete(createFullKeyById(id))
+                    .then(saveArtworkToRedis(savedArtwork))
+                    .thenReturn(savedArtwork)
+                    .onErrorResume(::isErrorFromRedis) { savedArtwork.toMono() }
```
Comment (ajax-gnatiuk-s) [inline domainservice/src/main/kotlin/ua/marchenko/artauction/artwork/repository/impl/RedisArtworkRepository.kt:35]: What if:

1. we saved a data to mongo and redis
2. we requested the data and received the cached value from redis
3. we updated the data in mongo and failed to update redisc
4. we requested the data and received the outdated cached value from redis

File: domainservice/src/main/kotlin/ua/marchenko/artauction/artwork/repository/impl/RedisArtworkRepository.kt (hunk 2 above)
Comment (ajax-marchenko-s) [reply domainservice/src/main/kotlin/ua/marchenko/artauction/artwork/repository/impl/RedisArtworkRepository.kt:31]: I have nullable id in `MongoArtwork`, so I need to check the id for nullability

File: domainservice/src/main/kotlin/ua/marchenko/artauction/artwork/repository/impl/RedisArtworkRepository.kt (hunk 3 above)
Comment (ajax-marchenko-s) [reply domainservice/src/main/kotlin/ua/marchenko/artauction/artwork/repository/impl/RedisArtworkRepository.kt:35]: User would receive the outdated values while the cache in redis exist (TTL 10 minutes in my case). Should I try to retry updating data in redis?

Comment (ajax-marchenko-s) [reply domainservice/src/main/kotlin/ua/marchenko/artauction/artwork/repository/impl/RedisArtworkRepository.kt]: I have added retries (5 with delay of 2 sec) to update keys in redis after saving/updating operations, but if redis is unavailable more than retries time, the user will recieve the outdated cached (while there are exist due to TTL)

**Review by ajax-gnatiuk-s — APPROVED** [review]

File: domainservice/src/main/kotlin/ua/marchenko/artauction/artwork/repository/impl/RedisArtworkRepository.kt (hunk 3 above)
Comment (ajax-marchenko-s) [reply domainservice/src/main/kotlin/ua/marchenko/artauction/artwork/repository/impl/RedisArtworkRepository.kt]: changed to backoff with 5 attempts and min delay of 200 millis

**Review by ajax-gnatiuk-s — APPROVED** [review]

**Pull request merged**